builder error: struct or union expected

### Describe the bug

When compiling the code it outputs a builder error.

### Reproduction Steps

```v
struct Parse {
mut:
        stack []&Element
}

struct Balise {}

struct RawText {}

type Element = Balise | RawText

fn (mut p Parse) process_open_tag() {
        mut last := &p.stack[0]
        if mut last is RawText {
        }
}
```

### Expected Behavior

No builder error or an error explaining what is wrong.

### Current Behavior

```
❯ v run main.v
==================
/tmp/v_1000/main.01J1506VYF091BH3R85WEJH7MS.tmp.c:13185: error: struct or union expected
...
==================
(Use `v -cg` to print the entire error message)

builder error
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.6 294f7e4.4d2c2da

### Environment details (OS name and version, etc.)

`OS: linux, Linux version 6.7.11-200.fc39.x86_64 (mockbuild@bdb8d24f206645e2bda44c1cf867612d) (gcc (GCC) 13.2.1 20240316 (Red Hat 13.2.1-7), GNU ld version 2.40-14.fc39) #1 SMP PREEMPT_DYNAMIC Wed Mar 27 16:50:39 UTC 2024`
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4931,7 +4931,14 @@ fn (mut g Gen) ident(node ast.Ident) {
 								if node.obj.is_inherited {
 									g.write(closure_ctx + '->')
 								}
-								g.write(name)
+								if node.obj.typ.nr_muls() > 1 {
+									g.write('(')
+									g.write('*'.repeat(node.obj.typ.nr_muls() - 1))
+									g.write(name)
+									g.write(')')
+								} else {
+									g.write(name)
+								}
 								if node.obj.orig_type.is_ptr() {
 									is_ptr = true
 								}

--- a/vlib/v/gen/c/infix.v
+++ b/vlib/v/gen/c/infix.v
@@ -695,6 +695,9 @@ fn (mut g Gen) infix_expr_is_op(node ast.InfixExpr) {
 
 	cmp_op := if node.op == .key_is { '==' } else { '!=' }
 	g.write('(')
+	if node.left_type.nr_muls() > 1 {
+		g.write('*'.repeat(node.left_type.nr_muls() - 1))
+	}
 	if is_aggregate {
 		g.write('${node.left}')
 	} else {

--- a/vlib/v/tests/sumtype_with_reference_test.v
+++ b/vlib/v/tests/sumtype_with_reference_test.v
@@ -1,0 +1,29 @@
+struct Parse {
+mut:
+	stack []&Element
+}
+
+struct Balise {}
+
+struct RawText {
+	s string
+}
+
+type Element = Balise | RawText
+
+fn (mut p Parse) process_open_tag() string {
+	mut last := &p.stack[0]
+	if mut last is RawText {
+		println(last)
+		return last.s
+	} else {
+		return ''
+	}
+}
+
+fn test_sumtype_with_reference() {
+	mut parse := Parse{
+		stack: [&RawText{'raw'}]
+	}
+	assert parse.process_open_tag() == 'raw'
+}

